File: SpaceDock/email.py
from SpaceDock.celery import send_mail
from SpaceDock.config import cfg
from werkzeug.utils import secure_filename

def send_confirmation(user, followMod=None):
    with open("emails/confirm-account") as f:
        confirmation = ""
        if followMod != None:
            confirmation = user.confirmation + "?f=" + followMod
        else:
            confirmation = user.confirmation
        message = f.read().format(
            **{ 
                'site_name': cfg['site-name'], 
                'username': user.username, 
                'domain': cfg['domain'],
                'confirmation': confirmation
            })
        send_mail.delay(cfg['support-mail'], [ user.email ], "Welcome to " + cfg['site-name'] + "!", message, important=True)

def send_reset(user):
    with open("emails/password-reset") as f:
        message = f.read().format(
            **{ 
                'site_name': cfg['site-name'], 
                'username': user.username, 
                'domain': cfg['domain'], 
                'confirmation': user.passwordReset 
             })
        send_mail.delay(cfg['support-mail'], [ user.email ], "Reset your password on " + cfg['site-name'], message, important=True)

def send_grant_notice(mod, user):
    with open("emails/grant-notice") as f:
        message = f.read().format(
            **{ 
                'username': user.username, 
                'mod_username': mod.user.username,
                'mod_name': mod.name,
                'site_name': cfg['site-name'], 
                "domain": cfg["domain"],
                'url': create_mod_url(mod.id, mod.name) 
            })
        send_mail.delay(cfg['support-mail'], [ user.email ], "You've been asked to co-author a mod on " + cfg['site-name'], message, important=True)

def send_update_notification(mod, version, user):
    followers = [u.email for u in mod.followers]
    changelog = version.changelog
    if changelog:
        changelog = '\n'.join(['    ' + l for l in changelog.split('\n')])

    targets = list()
    for follower in followers:
        targets.append(follower)
    if len(targets) == 0:
            return
    with open("emails/mod-updated") as f:
        message = f.read().format(
            **{
                'username': user.username,
                'friendly_version': version.friendly_version,
                'mod_name': mod.name,
                'site_name': cfg['site-name'],
                'changelog': changelog,
                'domain': cfg["domain"],
                'url': create_mod_url(mod.id, secure_filename(mod.name)[:64]),
                'game_name': version.gameversion.game.name,
                'gameversion': version.gameversion.friendly_version
            })
        subject = user.username + " has just updated " + mod.name + "!"
        send_mail.delay(cfg['support-mail'], targets, subject, message)

def send_autoupdate_notification(mod):
    followers = [u.email for u in mod.followers]
    changelog = mod.default_version().changelog
    if changelog:
        changelog = '\n'.join(['    ' + l for l in changelog.split('\n')])

    targets = list()
    for follower in followers:
        targets.append(follower)
    if len(targets) == 0:
        return
    with open("emails/mod-autoupdated") as f:
        message = f.read().format(
            **{
                'username': mod.user.username,
                'friendly_version': mod.default_version().friendly_version,
                'mod_name': mod.name,
                'game_name': mod.game.name,
                'gameversion': mod.default_version().gameversion.friendly_version,
                'domain': cfg["domain"],
                'url': create_mod_url(mod.id, secure_filename(mod.name)[:64])
            })
        subject = mod.name + " is compatible with " + mod.game.name + " " + mod.versions[0].gameversion.friendly_version + "!"
        send_mail.delay(cfg['support-mail'], targets, subject, message)

def send_bulk_email(users, subject, body):
    targets = list()
    for u in users:
        targets.append(u)
    send_mail.delay(cfg['support-mail'], targets, subject, body)

def create_mod_url(id, name):
    route = cfg['mod-url']
    return route.replace('{id}', str(id)).replace('{name}', name) # Using manual replacement here, so users dont need to use both values

File: utils/email.go
/*
 SpaceDock Backend
 API Backend for the SpaceDock infrastructure to host modfiles for various games

 SpaceDock Backend is licensed under the Terms of the MIT License.
 Copyright (c) 2017 Dorian Stoll (StollD), RockyTV
 */

package utils

import (
    "github.com/KSP-SpaceDock/SpaceDock-Backend/app"
    "github.com/go-gomail/gomail"
    "github.com/kennygrant/sanitize"
    "io/ioutil"
    "log"
    "strconv"
    "strings"
)

func SendMail(sender string, recipients []string, subject string, message string, important bool) {
    if app.Settings.SmtpHost == "" {
        return
    }
    srv := gomail.NewDialer(app.Settings.SmtpHost, app.Settings.SmtpPort, app.Settings.SmtpUser, app.Settings.SmtpPassword)
    m := gomail.NewMessage()
    if important {
        m.SetHeader("X-MC-Important", "true")
    }
    m.SetHeader("X-MC-PreserveRecipients", "false")
    m.SetHeader("Subject", subject)
    m.SetHeader("From", sender)
    if len(recipients) > 1 {
        m.SetHeader("Precedence", "bulk")
        m.SetHeader("To", "undisclosed-recipients:;")
    } else {
        m.SetHeader("To", recipients[0])
    }
    m.SetBody("text/plain", message)
    sc,err := srv.Dial()
    if err != nil {
        log.Printf("Error while sending mail: %s", err)
        return
    }
    defer sc.Close()
    sc.Send(sender, recipients, m)
    log.Printf("Sending email from %s to %d recipients", sender, len(recipients))
}

func SendConfirmation(userConfirmation string, userUsername string, userEmail string, followMod string) {
    buffer,err := ioutil.ReadFile("emails/confirm-account")
    if err != nil {
        log.Printf("Error while reading Email Template confirm-account: %s", err)
        return
    }
    confirmation := userConfirmation
    if followMod != "" {
        confirmation += "?f=" + followMod
    }
    data := map[string]interface{}{
        "site_name": app.Settings.SiteName,
        "username": userUsername,
        "domain": app.Settings.Domain,
        "confirmation": confirmation,
    }
    text := string(buffer)
    s := Format(text, data)
    go SendMail(app.Settings.SupportMail, []string{userEmail}, "Welcome to " + app.Settings.SiteName + "!", s, true)
}

func SendReset(userUsername string, userPasswordReset string, userEmail string) {
    buffer,err := ioutil.ReadFile("emails/password-reset")
    if err != nil {
        log.Printf("Error while reading Email Template password-reset: %s", err)
        return
    }
    data := map[string]interface{}{
        "site_name": app.Settings.SiteName,
        "username": userUsername,
        "domain": app.Settings.Domain,
        "confirmation": userPasswordReset,
    }
    text := string(buffer)
    s := Format(text, data)
    go SendMail(app.Settings.SupportMail, []string{userEmail}, "Reset your password on " + app.Settings.SiteName, s, true)
}

func SendGrantNotice(userUsername string, modUsername string, modName string, modID uint, userEmail string, modURL string) {
    buffer,err := ioutil.ReadFile("emails/grant-notice")
    if err != nil {
        log.Printf("Error while reading Email Template grant-notice: %s", err)
        return
    }
    data := map[string]interface{}{
        "username": userUsername,
        "mod_username": modUsername,
        "mod_name": modName,
        "site_name": app.Settings.SiteName,
        "domain": app.Settings.Domain,
        "url": create_mod_url(modID, sanitize.BaseName(modName)[:64], modURL),
    }
    text := string(buffer)
    s := Format(text, data)
    go SendMail(app.Settings.SupportMail, []string{userEmail}, "You've been asked to co-author a mod on " + app.Settings.SiteName, s, true)
}

func SendUpdateNotification(followers []string, changelog string, username string, friendly_version string, modname string, modID uint, modURL string, gamename string, gameversion string) {
    buffer,err := ioutil.ReadFile("emails/mod-updated")
    if err != nil {
        log.Printf("Error while reading Email Template mod-updated: %s", err)
        return
    }
    changelog = strings.Replace(changelog, "\n", "\n    ", -1)
    if len(followers) == 0 {
        return
    }
    data := map[string]interface{}{
        "username": username,
        "friendly_version": friendly_version,
        "mod_name": modname,
        "site_name": app.Settings.SiteName,
        "changelog": changelog,
        "domain": app.Settings.Domain,
        "url": create_mod_url(modID, sanitize.BaseName(modname)[:64], modURL),
        "game_name": gamename,
        "gameversion": gameversion,
    }
    text := string(buffer)
    s := Format(text, data)
    go SendMail(app.Settings.SupportMail, followers, username + " has just updated " + modname + "!", s, true)
}

func SendAutoUpdateNotification(followers []string, changelog string, username string, friendly_version string, modname string, modID uint, modURL string, gamename string, gameversion string) {
    buffer,err := ioutil.ReadFile("emails/mod-autoupdated")
    if err != nil {
        log.Printf("Error while reading Email Template mod-autoupdated: %s", err)
        return
    }
    changelog = strings.Replace(changelog, "\n", "\n    ", -1)
    if len(followers) == 0 {
        return
    }
    data := map[string]interface{}{
        "username": username,
        "friendly_version": friendly_version,
        "mod_name": modname,
        "game_name": gamename,
        "gameversion": gameversion,
        "domain": app.Settings.Domain,
        "url": create_mod_url(modID, sanitize.BaseName(modname)[:64], modURL),
    }
    text := string(buffer)
    s := Format(text, data)
    go SendMail(app.Settings.SupportMail, followers, modname + " is compatible with " + gamename + " " + gameversion + "!", s, true)
}

func create_mod_url(id uint, name string, modURL string) string {
    if modURL == "" {
        modURL = app.Settings.ModUrl
    }
    return strings.Replace(strings.Replace(modURL, "{id}", strconv.Itoa(int(id)), -1), "{name}", name, -1)
}